### DISCUSSION TITLE
Type checks in macros?

### DISCUSSION THREAD

**Discussion by krlmlr** [duckdb/duckdb#8599] [General]
Is there a way to implement macros that behave differently depending on the type of the argument(s)?

In the R example below, the `as.Date()` macro works for string inputs but fails for timestamp inputs. How would I go about implementing a version of this that works for both strings and timestamps?

(Trying to run this even with mainline duckdb will crash, you need https://github.com/duckdb/duckdb/compare/master...krlmlr:duckdb:b-altrep-unwind.)

``` r
con <- DBI::dbConnect(duckdb::duckdb())
experimental <- FALSE
invisible(DBI::dbExecute(con, "CREATE MACRO \"as.Date\"(x) AS strptime(x, '%Y-%m-%d')"))
df1 <- tibble::tibble(TRTEDTM = structure(1692244992, class = c("POSIXct", "POSIXt")))

rel1 <- duckdb:::rel_from_df(con, df1, experimental = experimental)
rel2 <- duckdb:::rel_project(
  rel1,
  list({
    tmp_expr <- duckdb:::expr_function("as.Date", list(duckdb:::expr_reference("TRTEDTM")))
    duckdb:::expr_set_alias(tmp_expr, "TRTEDTM")
    tmp_expr
  })
)
rel2
#> DuckDB Relation: 
#> ---------------------
#> --- Relation Tree ---
#> ---------------------
#> Projection [as.date(TRTEDTM) as TRTEDTM]
#>   r_dataframe_scan(0x10eef1838)
#> 
#> ---------------------
#> -- Result Columns  --
#> ---------------------
#> - TRTEDTM (TIMESTAMP)
duckdb:::rel_to_altrep(rel2)
#> Error in row.names.data.frame(x): Error evaluating duckdb query: Invalid Input Error: Could not parse string "2023-08-17 04:03:12" according to format specifier "%Y-%m-%d"
#> 2023-08-17 04:03:12
#>            ^
#> Error: Full specifier did not match: trailing characters
```

<sup>Created on 2023-08-17 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by krlmlr** [comment]
On a related note: Is there a `TRY()` function that would return a typed `NULL` on failure?

**Comment by hannes** [comment]
CC @Tmonster

**Comment by hannes** [comment]
@lnkuiper we may be able to add type annotations to macros and use that to allow overloading. But not so sure.

**Comment by krlmlr** [comment]
We agreed that C++-level functions in the R layer will be a better solution. Closing for now.
